### ISSUE TITLE
RPC Integration issue - GetBlock

### ISSUE THREAD

**Issue by patricklodder** [rosetta-dogecoin/rosetta-dogecoin#17]
Currently [GetBlock](https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/bitcoin/client.go#L352) fails, with the following message:

```
{"err": "invalid response: 500 Internal Server Error {\"result\":null,\"error\":{\"code\":-1,\"message\":\"JSON value is not a boolean as expected\"},\"id\":1}\n: error fetching block by hash <hash>: unable to get current block"}
```

This is likely due to a difference between the expected JSON-RPC method parameters and the [dogecoin implementation](https://github.com/dogecoin/dogecoin/blob/master/src/rpc/blockchain.cpp#L731).

This currently also prevents the Indexer process to successfully detect that dogecoind is online.

Could be an easy fix?

**Comment by xanimo** [comment]
I can work on this one but may need some help along the way.

**Comment by patricklodder** [comment]
Anything you need :-)

**Comment by rllola** [comment]
Easy fix would be to remove the `blockVerbosity` parameter here https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/bitcoin/client.go#L365 ? If I am not mistaken by default in dogecoin this value is a boolean and it is true ?

But I guess we want to modify the files in the `bitcoin` folder as little as possible ?

I did try it. But I am having some issue with the docker build caching...

**Comment by xanimo** [comment]
This is my first foray into golang and I've been mostly learning about package management past 24 hr.  

It seems that in the docker build process we are still pulling a lot of files from coinbase/rosetta-bitcoin and changes made to local files are not being included in the build.  An example for verification purposes changed blockchain: Bitcoin to blockchain: Dogecoin in bitcoin.json, only to test if changes are reflected when config is loaded on boot and it is unfortunately not included.

I know this is per your recommendation for least amount of work in maintaining this repo and keeping it synced w upstream but I think we will have to at some point delineate for development purposes (I am thinking maybe a new branch using either a proxy repo of rosetta-bitcoin as a new intermediary temp upstream or looking into GOPROXY for local dev env).  

I know this is 'hacky' but typically the way I learn best is by making tweaks on my local to show more explicit errors in console logs and without the ability to make local changes and view the exact error, it feels like searching for a needle in a haystack.  For instance I have not been able to replicate that RPC error.  If you have time could you show me how to reproduce?  Or where to look for that in logs?  Let me know if you can think of something that might resolve this.

Outside of that, I encountered a new odd but trivial bug within WSL2 dev env wherein the linux subsystem and Windows clocks become unsynced and dockerd complains about systems date & time being set correctly.  This was easily solved by entering sudo hwclock --htcosys which syncs subsystem time with Windows.  Only applicable to folks using WSL2.

Anyway hope all are doing well!  Still plugging away. ❤️

**Comment by patricklodder** [comment]
> Can you provide steps to reproduce or where to look for logs?

see https://gist.github.com/patricklodder/f47a1e5fd03ddaf1e29a83c81c181edf

After downloading, you can apply the patch with `git apply add_net_status_probe.patch` on current main branch.

**Comment by RichDevX** [comment]
Assigning a boolean value to blockVerbosity instead of an integer should fix the issue.

**Comment by xanimo** [comment]
> > Can you provide steps to reproduce or where to look for logs?
> 
> see https://gist.github.com/patricklodder/f47a1e5fd03ddaf1e29a83c81c181edf
> 
> After downloading, you can apply the patch with `git apply add_net_status_probe.patch` on current main branch.

Great thank you!  Will fix this shortly.

**Comment by xanimo** [comment]
I apologize for the delay but I am still having troubles with what seems like it should be a quick fix.  Here are my logs:

`panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x94665b]

goroutine 11 [running]:
github.com/coinbase/rosetta-bitcoin/bitcoin.(*Client).post(0x0, 0xe70af8, 0xc00003ad00, 0xc74762, 0x11, 0x127a680, 0x0, 0x0, 0xe64060, 0xc000180b80, ...)
        /home/xanimo/go/pkg/mod/github.com/coinbase/rosetta-bitcoin@v0.0.9/bitcoin/client.go:825 +0x1fb
github.com/coinbase/rosetta-bitcoin/bitcoin.(*Client).getBlockchainInfo(0x0, 0xe70af8, 0xc00003ad00, 0x0, 0x0, 0xc0005555c0)
        /home/xanimo/go/pkg/mod/github.com/coinbase/rosetta-bitcoin@v0.0.9/bitcoin/client.go:381 +0xa5
github.com/coinbase/rosetta-bitcoin/bitcoin.(*Client).getBlockHash(0x0, 0xe70af8, 0xc00003ad00, 0x0, 0xc00055575c, 0x2, 0x2, 0x10100000000)
        /home/xanimo/go/pkg/mod/github.com/coinbase/rosetta-bitcoin@v0.0.9/bitcoin/client.go:398 +0x107
github.com/coinbase/rosetta-bitcoin/bitcoin.(*Client).getBlock(0x0, 0xe70af8, 0xc00003ad00, 0x0, 0xc00055575c, 0xc00008b038, 0xc00055575c)
        /home/xanimo/go/pkg/mod/github.com/coinbase/rosetta-bitcoin@v0.0.9/bitcoin/client.go:356 +0x67
github.com/coinbase/rosetta-bitcoin/bitcoin.(*Client).NetworkStatus(0x0, 0xe70af8, 0xc00003ad00, 0x0, 0x2, 0x1)
        /home/xanimo/go/pkg/mod/github.com/coinbase/rosetta-bitcoin@v0.0.9/bitcoin/client.go:166 +0x55
main.main.func5(0xc0001ea410, 0x0, 0xe70af8, 0xc00003ad00, 0xc0000121a0, 0xc00001e5a0)
        /home/xanimo/source/repos/rosetta-dogecoin/main.go:199 +0x151
created by main.main
        /home/xanimo/source/repos/rosetta-dogecoin/main.go:195 +0xa37`

Does this mean that changes made to 'our' bitcoin/client.go are irrelevant or do they merge directories on build or someting?  If so how can we point to use our client.go?

**Comment by patricklodder** [comment]
@xanimo can you do a draft PR of where you are at so that we can see the code, do a checkout and reproduce locally? That way it's easier to support.

**Comment by xanimo** [comment]
> @xanimo can you do a draft PR of where you are at so that we can see the code, do a checkout and reproduce locally? That way it's easier to support.

It's the most bare bones PR yet haha.  But yeah I'm still having trouble going to call it a night!

**Comment by RichDevX** [comment]
An intermediate Block structure (defined in bitcoin/types.go) will need to be created, getblock "blockhash" true/1 returns an array of transaction id strings, whereas getblock "blockhash" 2 returns an array of transaction objects. GetBlock will need to send a getrawtransaction request for each transaction id in the array to get the transaction data required to populate the Txs slice in the Block structure.

**Comment by RichDevX** [comment]
getrawtransaction is unable to retrieve the genesis block coinbase transaction, we would have to hard code a response, which isn't ideal. Alternatively we could decode the raw block data (getblock "blockhash" false) and use the decoded data to populate block and transaction structures. Until a new dogecoin core with an updated rpc interface is released, using the raw block data seems to be the most efficient solution.

**Comment by victorsk2019** [comment]
> getrawtransaction is unable to retrieve the genesis block coinbase transaction

Hi.  Could you post the error stack trace you are getting please?  Thanks.

**Comment by xanimo** [comment]
> > getrawtransaction is unable to retrieve the genesis block coinbase transaction
> 
> Hi.  Could you post the error stack trace you are getting please?  Thanks.

Think we got it actually. 🙂

**Comment by victorsk2019** [comment]
> > > getrawtransaction is unable to retrieve the genesis block coinbase transaction
> > 
> > 
> > Hi.  Could you post the error stack trace you are getting please?  Thanks.
> 
> Think we got it actually. 🙂

I have no doubt of that.  I was just curious to see what that error looks like so I could compare it with errors I get and hopefully learn more about the project :)

**Comment by xanimo** [comment]
> > > > getrawtransaction is unable to retrieve the genesis block coinbase transaction
> > > 
> > > 
> > > Hi.  Could you post the error stack trace you are getting please?  Thanks.
> > 
> > 
> > Think we got it actually. 🙂
> 
> I have no doubt of that. I was just curious to see what that error looks like so I could compare it with errors I get and hopefully learn more about the project :)

This was the getrawtransaction error:

![Screenshot (380)](https://user-images.githubusercontent.com/31856436/113451853-b468d700-93b7-11eb-9f9d-a1f7ab2d37ed.png)

**Comment by JustInTrades** [comment]
Hi! Just wondering given the current status an estimate of finishing coding?

**Comment by RichDevX** [comment]
@victorsk2019 the genesis block is hard coded, so by design getrawtransaction is unable to get the coinbase transaction from the genesis block. What we did was use getblock "hash" false to get the raw blocks which includes the transaction data/bytes, and then used decoderawtransaction to decode each transaction. It's definitely not the most efficient solution, but I figure the next dogecoin core release will have an updated rpc interface, and the solution will no longer be necessary.

**Comment by xanimo** [comment]
> @victorsk2019 the genesis block is hard coded, so by design getrawtransaction is unable to get the coinbase transaction from the genesis block. What we did was use getblock "hash" false to get the raw blocks which includes the transaction data/bytes, and then used decoderawtransaction to decode each transaction. It's definitely not the most efficient solution, but I figure the next dogecoin core release will have an updated rpc interface, and the solution will no longer be necessary.

Credit to @RichDevX for getBlock func decoding the genesis block transaction lines 383-453 in dogecoin/client.go.  Different variables in my branch but same function.

And on intermediary block struct to hold raw transaction in string array before decoding and splicing into *Transaction array in Block struct response.  Lines 224-269 and 388-408 in dogecoin/types.go.

**Comment by victorsk2019** [comment]
> > @victorsk2019 the genesis block is hard coded, so by design getrawtransaction is unable to get the coinbase transaction from the genesis block. What we did was use getblock "hash" false to get the raw blocks which includes the transaction data/bytes, and then used decoderawtransaction to decode each transaction. It's definitely not the most efficient solution, but I figure the next dogecoin core release will have an updated rpc interface, and the solution will no longer be necessary.
> 
> Credit to @RichDevX for getBlock func decoding the genesis block transaction lines 383-453 in dogecoin/client.go. Different variables in my branch but same function.
> 
> And on intermediary block struct to hold raw transaction in string array before decoding and splicing into *Transaction array in Block struct response. Lines 224-269 and 388-408 in dogecoin/types.go.

Thank you for clarifying.  It makes it easier for me to spot changes rather than comparing code line-by-line :)

**Comment by patricklodder** [comment]
Fixed for v1 blocks with #19
